### PR TITLE
Fix API inconsistency of NULL Date

### PR DESCRIPTION
**Pull request by thbeu**
It is possible to write a NULL Date by `DBFWriteNULLAttribute` where a subsequent call to `DBFIsAttributeNULL` returns false due to handling only `"0"` and `"00000000"` as valid NULL Date strings. This is fixed by writing NULL Date independently of the field width.

Edit: There are other fixes possible, e.g. for width of 1 and 8 to be more backward-compatible w.r.t. binary content or by considering the field width when reading the Null string.

### CODE REVIEW

File: dbfopen.c [hunk 1]
```diff
@@ -1311,9 +1311,18 @@ static bool DBFWriteAttribute(DBFHandle psDBF, int hEntity, int iField,
...
+            // NULL Date is independent of panFieldSize
+            pabyRec += psDBF->panFieldOffset[iField];
+            pabyRec = SHPLIB_NULLPTR;
+        }
+        else
+        {
+            memset(pabyRec + psDBF->panFieldOffset[iField],
+                   DBFGetNullCharacter(psDBF->pachFieldType[iField]),
+                   psDBF->panFieldSize[iField]);
+        }
```
Comment (rouault) [inline dbfopen.c:1325]: This change looks completely wrong to me

Comment (thbeu) [reply dbfopen.c:1325]: As I said, other fixes are possible to get the failing tests green.

Comment (rouault) [reply dbfopen.c:1325]: I don't understand why it is necessary to change that part of the code. What's wrong with the current code? As far as I understand, it should fill the date field with the '0' character, which is the expected behavior

Comment (thbeu) [reply dbfopen.c:1325]: No, `DBFWriteNULLAttribute` writes a string of  "0" repeated `panFieldSize` times, but only "0" and "00000000" is considered NULL Date in `DBFIsAttributeNULL `.

Comment (rouault) [reply dbfopen.c:1325]: > `DBFWriteNULLAttribute` writes a string of "0" repeated `panFieldSize` times

That should be fine. panFieldSize[] should be 8 for a date field

Comment (thbeu) [reply dbfopen.c:1325]: > That should be fine. panFieldSize[] should be 8 for a date field

Which is not checked or verified. `DBFWriteNULLAttribute` could be fixed to return false on unexpected `panFieldSize`.

Comment (rouault) [reply dbfopen.c:1325]: wouldn't a better fix to check that the size is 8 for date fields in DBFAddNativeFieldType ?

Comment (thbeu) [reply dbfopen.c:1325]: Sounds very reasonable to me. However I am not sure if it breaks existing applications. Should we also allow size = 1 (such that " " and "0" can be successfully read as NULL Date strings?

Comment (rouault) [reply dbfopen.c:1325]: There's indeed a slight breakage possibility, but I've skimmed a bit various DBF documentation, and all document Date fields being 8 byte large, so anyone producing such files would produce unusual/non-conformant files. If we keep that restriction on DBFAddNativeFieldType(), we still allow reading those unusual/non-conformant files

The alternatives would be:
- not care about DBFIsAttributeNULL() not working for date field of size != 8
- or make DBFIsAttributeNULL() tolerant to be set to as many '0' as the field width

Comment (thbeu) [reply dbfopen.c]: I went for the second alternative and updated this PR accordingly.

**Pull request merged**